### ISSUE TITLE
About unlabeled_vae_loss

### ISSUE THREAD

**Issue by csyanbin** [bjlkeng/sandbox#3]
Hi, bjlkeng,   
     I read your blog and GitHub code. That's quite helpful for me as a new learner to VAE methods.

     When I was reading the code of semi-VAE m2 model, I found that the unlabeled_vae_loss function was different from Kingma's paper?
     I am not quite sure, but should it be an iteration over all possible y and a sum loss as in the first term of Eq(7) in paper "Semi-supervised Learning with Deep Generative Models"?
     
     I will appreciate it if you can reply to my confusion.

**Comment by bjlkeng** [comment]
Hi @csyanbin, good question!  Had to spend a few minutes refreshing my memory of what I did but here's my interpretation:

For Equation 7 from the original paper you have:
```
[ \sum_y q(y|x)(-L(x,y)) ] + H(q(y|x))
```
We'll only focus on the first term since that's where your question is I think.

If we look at `L(x,y)`, we see that it's dependence on `y` is only from the decoder output `log p(x|y,z)`.  But you'll notice that for any given training sample `x_0` and random latent variable sample `z_0`, this produces exactly the same reconstructed output.  This is because `y` is constant with respect to a given `x_0` since it's computed from the discriminator network.  

So looking back at the first term above, we get for a given `x_0` and `z_0` sample:

```
\sum_y q(y|x)(-L(x,y)) = \sum_y q(y|x_0) f(x_0, z_0) = f(x_0, z_0) \sum_y q(y|x_0) = f(x_0, z_0)
```
The last equality is from the fact that `y` is a categorical distribution so it sums to 1.  So essentially it washes out.

-----
I was thinking about a different way to go about it though.  For a given `x_0`, you could try to sample `y` from the output of the discriminator network (based on the distribution) and then try to do the `q(y|x)` weighted summation as suggest but I don't really think that would work out to the same thing.

Let me know if this makes sense, I could have some flaw in my logic.  Hope this helps.

**Comment by csyanbin** [comment]
Hi @bjlkeng,

Many thanks for the fast response and really learned much from your VAE related blogs.

I can understand your explanation. But I am still confused about why "This is because y is constant with respect to a given x_0 since it's computed from the discriminator network."
     I read other implementations of semi-AVE, e.g. https://github.com/saemundsson/semisupervised_vae/blob/master/genclass.py#L201. They just integrating out y for unlabeled data by a for loop over possible labels.   

For the sampling of y, I think maybe Gumbel softmax algorithm is helpful, e.g., in Section 2.4 of "Improved Variational Autoencoders for Text Modeling using Dilated Convolutions".   

Thanks so much for your answer! I am very glad to communicate and discuss with you.

**Comment by bjlkeng** [comment]
Hi @csyanbin, I think this is correct.  What I did kind of looks like a hack, that now looking back, I'm a bit surprised it works.

Let's say you have 3 categories for `y`.  Basically, for each training example, you'll have to put `{[1, 0, 0], [0, 1, 0], [0, 0, 1]`, through the decoder as the `y` value, and compute the gradient 3 times.  I did something simpler which is the output of the discriminator gives a distribution across `y` values, so I just put that (e.g. `[0.1, 0.5, 0.4]`) into the decoder.  The upside is that you only need to do one pass of the decoder network but at the same time, it's not computing what the paper did.

Thanks for bringing this up!  I'll put a not in the notebooks and blog linking back to this discussion, although I probably won't bother re-running them.

**Comment by csyanbin** [comment]
Hi @bjlkeng,

Many thanks for the explanation, I think your trick seems reasonable and actually saved a lot of computation time. 

At the same time, I think some techniques of sampling from discrete distributions, such as Gumbel softmax is also a good solution for reducing computations. 
For your reference, I just encountered this implementation: https://github.com/ryokamoi/dcnn_textvae/blob/master/model/discriminator.py#L50, which implemented Gumbel in Section 2.4 of "Improved Variational Autoencoders for Text Modeling using Dilated Convolutions".

Hope this will be helpful for you!

**Comment by bjlkeng** [comment]
@csyanbin thanks for the link!  I'll definitely take a look at it.  Wish I had more time to explore some of these ideas but it's hard to do it when you're not a full time academic!  Good luck with your project.

**Comment by wzm2256** [comment]
Hi, to make things clear, I read the original code of the author https://github.com/dpkingma/nips14-ssl.git , and find the commont in 254 line in the file learn_yz_x_ss.py:

        # === Get gradient for unlabeled data
        # -KL(q(z|x,y)q(y|x) ~p(x) || p(x,y,z))
        # Approach where outer expectation (over q(z|x,y)) is taken as explicit sum (instead of sampling)

The logic is very close to the tensorflow version in the link
https://github.com/saemundsson/semisupervised_vae/blob/master/genclass.py#L201


So, I'm quite sure it's better to sum out the expectation term in the unsupervised term,
However, this implement could be horrible if the number of class is large, say for example 1000 in ImageNet.

**Comment by bjlkeng** [comment]
Thanks for the clarification @wzm2256 !  Definitely makes more sense than what I implemented.  As you point out though, it could be very computationally expensive.  Hopefully this thread can serve to help those who are confused by this in the future.
